### PR TITLE
BUGFIX: Nodes are inaccessible after base workspace switch

### LINKED ISSUE #1602: Nodes are inaccessible after base workspace switch
A problem with the routing cache may result in inaccessible document nodes for cases where nodes with different identifiers but the same URI path exist in separate workspaces.

Steps to reproduce:

- create two workspaces ("A" and "B") which are based on "live"
- create a document node with uri path segment "foo" and publish it to live
- switch to workspace "B"
- change the title of the page, publish it to workspace "B"
- switch to workspace "A"
- delete the document node "foo", publish it to workspace "A"
- publish all changes of workspace "A" to workspace "live"
- create a new document node with the same name "foo", publish it to workspace "A"
- switch to workspace "B"

You will now get a routing error because the cached route for "/foo" points to the node identifier of the node in workspace "A" which does not exist in workspace "B".

- flush the routing cache

The document node "foo" can now be access in workspace "B"

- switch to workspace "A"

You will see the routing error again because now the cached route for "/foo" points to the node identifier of the node in workspace "B" which does not exist in workspace "A".

### PR DESCRIPTION
**Pull request by robertlemke**
This change fixes a problem with the routing cache which results in
inaccessible document nodes for cases where nodes with different
identifiers but the same URI path exist in separate workspaces.

Fixes #1602

### CODE REVIEW

File: TYPO3.Neos/Classes/TYPO3/Neos/Routing/Cache/RouteCacheFlusher.php [hunk 1]
```diff
@@ -52,6 +53,24 @@ public function registerNodeChange(NodeInterface $node)
...
+     * In most cases $workspace will be a user's personal workspace. Flushing the respective cache entries guards
+     * against mismatches for nodes which exist in the old and the new base workspace but have different node
+     * identifiers and the same URI path (segment).
+     *
+     * @param Workspace $workspace
+     * @param Workspace|null $oldBaseWorkspace
+     * @param Workspace|null $newBaseWorkspace
+     * @return void
+     */
+    public function registerBaseWorkspaceChange(Workspace $workspace, Workspace $oldBaseWorkspace = null, Workspace $newBaseWorkspace = null)
```
Comment (aertmann) [inline TYPO3.Neos/Classes/TYPO3/Neos/Routing/Cache/RouteCacheFlusher.php]: why the unused parameters?

Comment (robertlemke) [inline TYPO3.Neos/Classes/TYPO3/Neos/Routing/Cache/RouteCacheFlusher.php]: These arguments are provided by the respective emit method - I don't know what's more messy, to not use them in the slot or to have a method signature which does not comply with "the interface" of the signal.

I don't mind removing the arguments, but I think it's not better or worse this or the other way.

**Comment by kdambekalns** [comment]
I followed the instructions in the ticket, and could reproduce the error. Then I

- discarded the new node in A
- switched applied the PR
- created foo again in A (last but one step)

Now this gives me a 404 ("An error occurred" in the UI). Should that already have worked?

Anyway, after a flushing of the cache switching works as expected, _but_ the publish button flashes red for a moment before the page loads. And the document tree does not load, the "globe" of the root node stays a loading spinner… This is the browser console:

console-errors

**Comment by robertlemke** [comment]
I've seen that "Conflict" error, too. There's probably some more work to be done to provide a smooth workspace switch in these cases, so this fix here can only be seen as a remedy for the most problematic part, which is the deadlock of one document not being accessible at all.

My guess is that this needs to be tackled on the UI side as well.

**Comment by robertlemke** [comment]
Hmm, could it be that the UI loads the children before the cache is actually flushed? I'll try something …

**Comment by robertlemke** [comment]
Okay, for the record: In the `RouteCacheFlusher` I tried to flush the routing caches immediately when the base workspace changed, instead of delaying it until `shutdownObject`. But that didn't help either with the UI problem.

**Pull request merged**